### DISCUSSION TITLE
spdlog for different targets in one CMake project

### DISCUSSION THREAD

**Discussion by rroguski** [gabime/spdlog#2213] [Q&A]
Hello,

I've been using spdlog for quite some time now. I added it as a submodule to my project over 2 years ago, when in the README it still said to copy `src/spdlog.cpp` to project's source directory for faster compile times.

Recently I decided to update the submodule and after that had lots of linker errors. To make the long story short it turned out that templates are not initialized, because more files appeared next to `src/spdlog.cpp`. Then I noticed that the info on adding `src/spdlog.cpp` to my source directory is gone from the README and only the precompiled info remains.

Here is my problem - my project is cross-compiled for ARM with unit tests for x86_64 and both binaries use spdlog (since some tested modules require spdlog). The problem is that I would have to use 2 separately built spdlog libraries, but I use spdlog as a submodule in my project already so I'd rather have the old option with adding spdlog files to the rest of my compiled files.

I didn't want to add all files from `spdlog/src` by hand since with later updates when more files are added I'd have to make changes, so I did the following:

1. In my main `CMakeLists.txt` file I added the following line: `AUX_SOURCE_DIRECTORY(${CMAKE_SOURCE_DIR}/spdlog/src SPDLOG_COMPILE_FILES)`
2. In each `CMakeLists.txt` file which has an `ADD_EXECUTABLE` I added to the end of the file list `${SPDLOG_COMPILE_FILES}`

What `AUX_SOURCE_DIRECTORY` does is it saves the file names located in `spdlog/src` directory to `SPDLOG_COMPILE_FILES` variable.

My question is - is this ok? It seems to be update-proof since if new files are added the list of files updates automatically, but I have an impression that I'm doing something completely wrong and nobody is doing it this way since I can't find any information about it...

Thank you in advance for any insight on this problem.

**Comment by tt4g** [comment]
Usually, it is enough to call `add_subdirectory()` to include third party libraries.
CMake targets (e.g. spdlog) declared in `CMakeLists.txt` loaded by `add_subdirectory()` can be linked to targets in your project.

**Reply by rroguski** [reply]
Thanks for answering.

The thing is I have my main `CMakeLists.txt` which has many `ADD_SUBDIRECTORY` entries, which also include:
```
ADD_SUBDIRECTORY(spdlog)
ADD_SUBDIRECTORY(tests)
```
in `tests` I have another `CMakeLists.txt` which also requires `ADD_SUBDIRECTORY(spdlog)` and this causes a cmake error saying `add_library cannot create target "spdlog" because another target with the same name already exists.` (and the target exists because of the `add_subdirectory(spdlog)` in my main `CMakeLists.txt`).

I have a different library which is built normally for the main project and is additionally mocked for tests, but there, instead of a regular library name in `add_library`, I have a cmake variable set in main `CMakeLists.txt` to its regular library name and in the tests `CMakeLists.txt` the variable is set to a mock name.

Maybe an example will be more clear:
my library `CMakeLists.txt`:
```
ADD_LIBRARY(${LIBRARY_NAME}
   # cpp files...
)
```
main `CMakeLists.txt`:
```
# compiler for arm
SET(CMAKE_CXX_COMPILER "${MY_PATH_TO_COMPILER}/bin/arm-linux-gnueabihf-g++")
PROJECT(myproject)
# setting library name so there is no name clash
SET(LIBRARY_NAME mylib)
ADD_SUBDIRECTORY(mylib)
ADD_SUBDIRECTORY(spdlog)
ADD_SUBDIRECTORY(tests)
# ...
```
tests `CMakeLists.txt`:
```
# compiler for x86_64
SET(CMAKE_CXX_COMPILER "/usr/bin/g++")
PROJECT(tests)
# setting library name so there is no name clash
SET(LIBRARY_NAME mylibMock)
ADD_SUBDIRECTORY(mylib)
ADD_SUBDIRECTORY(spdlog)
# ...
```

No name clash on mylib since for the main project it is linked as `mylib` and for tests it is linked as `mylibMock` (`mylib` (linked with ARM-specific stuff, `mylibMock` linked with dummy ARM-specific stuff, not relevant but just for clarification). With spdlog there are 2  same names, hence the error.

I think my main problem is that I'm building 2 separate binaries for 2 separate targets in one go. This might be the reason why no one does it this way :wink:

**Reply by tt4g** [reply]
Right. If you import the same project twice with `add_subdirectory()`, the CMake target names will conflict and should be avoided.
Remove `add_subdirectory(spdlog)` in `tests` because spdlog is already defined in `main`.

Use this small sample project as a reference: https://github.com/tt4g/spdlog-discussions-2213

**Reply by rroguski** [reply]
Thank you for the example. It works great when `spdlog_discussions_2213` and `spdlog_discussions_2213_test` are compiled for the same target. `spdlog` is built as a static library and linked to both binaries.

But if I change the compiler for `spdlog_discussions_2213` to a cross-compiler for ARM and set a compiler for `spdlog_discussions_2213_test` to my default compiler on my PC then the statically built `spdlog` library (built using the cross-compiler for ARM) is being linked corerctly to `spdlog_discussions_2213`, but there in a linker error when it's trying to link to the `spdlog_discussions_2213_test` binary built for a different target.

I think you're missing the fact that I'm creating one binary for ARM and another for x86_64. Adding `spdlog` as a subdirectory in one place results in the `spdlog` static library being built for only one of those targets. I can't add spdlog twice using `add_subdirectory` because of the name clash. That is why I had to go back to the option of adding `spdlog` sources to my build tree (this option was deleted in commit https://github.com/gabime/spdlog/commit/f0f4499540484dec7febed0e3b599182e6012dae).

So the question is why was the option with adding `cpp` files to the build tree removed from the description?

**Reply by tt4g** [reply]
> So the question is why was the option with adding cpp files to the build tree removed from the description?

I can't answer this question. @gabime Could you please answer it?
